spec/unit/AndroidProject.spec.js

```diff
@@ -0,0 +1,134 @@
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+*/
+
+const path = require('path');
+const rewire = require('rewire');
+
+describe('AndroidProject', () => {
+    const PROJECT_DIR = 'platforms/android';
+    let AndroidProject;
+    let AndroidStudioSpy;
+
+    beforeEach(() => {
+        AndroidProject = rewire('../../bin/templates/cordova/lib/AndroidProject');
+
+        AndroidStudioSpy = jasmine.createSpyObj('AndroidStudio', ['isAndroidStudioProject']);
+        AndroidProject.__set__('AndroidStudio', AndroidStudioSpy);
+    });
+
+    describe('constructor', () => {
+        it('should set the project directory and www folder', () => {
+            const project = new AndroidProject(PROJECT_DIR);
+            expect(project.projectDir).toBe(PROJECT_DIR);
+        });
+
+        it('should set www folder correctly if not Android Studio project', () => {
+            AndroidStudioSpy.isAndroidStudioProject.and.returnValue(false);
+
+            const project = new AndroidProject(PROJECT_DIR);
+            expect(project.www).toBe(path.join(PROJECT_DIR, 'assets/www'));
+        });
+
+        it('should set www folder correctly if it is an Android Studio project', () => {
+            AndroidStudioSpy.isAndroidStudioProject.and.returnValue(true);
+
+            const project = new AndroidProject(PROJECT_DIR);
+            expect(project.www).toBe(path.join(PROJECT_DIR, 'app/src/main/assets/www'));
+        });
+    });
+
+    describe('getProjectFile', () => {
+        it('should create and return a new project if one does not exist', () => {
+            const newProject = AndroidProject.getProjectFile(PROJECT_DIR);
+
+            expect(newProject).toEqual(jasmine.any(AndroidProject));
+        });
+
+        it('should return the existing project', () => {
+            const newProject = AndroidProject.getProjectFile(PROJECT_DIR);
+            const secondProject = AndroidProject.getProjectFile(PROJECT_DIR);
+
+            expect(newProject).toEqual(jasmine.any(AndroidProject));
+            expect(secondProject).toBe(newProject);
+        });
+    });
+
+    describe('purgeCache', () => {
+        beforeEach(() => {
+            AndroidProject.__set__('projectFileCache', {
+                project1: 'test',
+                project2: 'anothertest',
+                project3: 'finaltest'
+            });
+        });
+
+        it('should only remove the specified project from the cache', () => {
+            const projectToRemove = 'project2';
+            AndroidProject.purgeCache(projectToRemove);
+
+            const cache = AndroidProject.__get__('projectFileCache');
+            expect(Object.keys(cache).length).toBe(2);
+            expect(cache[projectToRemove]).toBe(undefined);
```

Comment:
I find `in` to be vastly underused in the JS community :wink:

Reply:
Yeah, it wouldn't even cross my mind to use it!